knpu/episode3/Challenges/PdoShipStorage/DecomposeStringTransformerCoding.php

```diff
@@ -0,0 +1,189 @@
+"Ha!" John says, "This is simple!". To show off, John creates the `StringTransformer`
+class and *even* makes it cache the results to be super-performant.
+
+But wait you say! Combining the string transformation *and* caching into the same
+class make `StringTransformer` responsible for two jobs. Help show Bob the intern
+a better way, by creating a new `Cache` class with two methods `fetchFromCache(\$key)`
+and `saveToCache(\$key, \$val)`. Then, pass this into `StringTransformer` and use it
+to cache, instead of using your own logic:
+EOF;
+    }
+
+    public function getFileBuilder()
+    {
+        $fileBuilder = new FileBuilder();
+        $fileBuilder
+            ->addFileContents('Cache.php', <<<EOF
+EOF
+            )
+            ->addFileContents('StringTransformer.php', <<<EOF
+<?php
+
+class StringTransformer
+{
+    public function transformString(\$str)
+    {
+        \$cacheFile = __DIR__.'/cache/'.md5(\$str);
+
+        if (file_exists(\$cacheFile)) {
+            return file_get_contents(\$cacheFile);
+        }
+
+        \$newStr = '';
+        foreach (str_split(strrev(\$str), 2) as \$twoChars) {
+            var_dump(\$twoChars);
+            // capitalize the first of 2 characters
+            \$newStr .= ucfirst(\$twoChars);
+        }
+
+        if (!file_exists(dirname(\$cacheFile))) {
+            mkdir(dirname(\$cacheFile));
+        }
+        file_put_contents(\$cacheFile, \$newStr);
+
+        return \$newStr;
+    }
+}
+EOF
+            )
+            ->addFileContents('index.php', <<<EOF
+<?php
+
+require 'Cache.php';
+require 'StringTransformer.php';
+
+\$str = 'Judge me by my size, do you?';
+
+\$transformer = new StringTransformer();
+var_dump(\$transformer->transformString(\$str));
+EOF
+            )
+            ->setEntryPointFilename('index.php')
+        ;
+
+        return $fileBuilder;
+    }
+
+    public function getExecutionMode()
+    {
+        return self::EXECUTION_MODE_PHP_NORMAL;
+    }
+
+    public function setupContext(CodingContext $context)
+    {
+    }
+
+    public function grade(CodingExecutionResult $result)
+    {
+        if (!class_exists('\Cache')) {
+            throw new GradingException('Class `Cache` does not exist. Did you create it?');
+        }
+        $cacheClass = new \ReflectionClass('\Cache');
+        if (!$cacheClass->hasMethod('fetchFromCache')) {
+            throw new GradingException('Method `fetchFromCache` does not exist in the `Cache` class.');
+        }
+        if (!$cacheClass->hasMethod('saveToCache')) {
+            throw new GradingException('Method `saveToCache` does not exist in the `Cache` class.');
+        }
+    }
+
+    public function configureCorrectAnswer(CorrectAnswer $correctAnswer)
+    {
+        $correctAnswer
+            ->setFileContents('Cache.php', <<<EOF
+<?php
+
+class Cache
+{
+    public function fetchFromCache(\$key)
+    {
+        \$cacheFile = __DIR__.'/cache/'.md5(\$key);
```

Comment:
Good idea - I added a private method actually to do this. It's a nice extra thing to show